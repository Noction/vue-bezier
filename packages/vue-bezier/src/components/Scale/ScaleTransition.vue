<template>
  <component
    :is="componentType"
    :tag="props.tag"
    name="noc-scale"
    v-bind="{ ...$attrs, ...hooks }"
  >
    <slot />
  </component>
</template>

<script setup lang="ts">
import type { ComponentEvents, ComponentProps } from '@/types'
import { buildComponentType, useHooks } from '@/composables'

defineOptions({ inheritAttrs: false })

const props = withDefaults(defineProps<ComponentProps>(), {
  delay: 0,
  duration: 300,
  group: false,
  origin: 'top left',
  styles: () => ({
    animationFillMode: 'both',
    animationTimingFunction: 'ease-out'
  }),
  tag: 'span'
})

const emit = defineEmits<ComponentEvents>()

const componentType = buildComponentType(props)

const hooks = useHooks(props, emit)

</script>

<style>
  @keyframes scale-in {
    from {
      opacity: 0;
      transform: scale(0);
    }

    to { opacity: 1; }
  }

  @keyframes scale-out {
    from { opacity: 1; }

    to {
      opacity: 0;
      transform: scale(0);
    }
  }

  .noc-scale-enter-active { animation-name: scale-in; }
  .noc-scale-leave-active { animation-name: scale-out; }
  .noc-scale-move { transition: transform .3s cubic-bezier(.25, .8, .5, 1); }
</style>
