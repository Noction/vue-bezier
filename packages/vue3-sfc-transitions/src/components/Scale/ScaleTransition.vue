<template>
  <component
    :is="componentType"
    :tag="tag"
    v-bind="{ ...$attrs, ...hooks }"
    enter-active-class="scale-in"
    move-class="scale-move"
    leave-active-class="scale-out"
  >
    <slot />
  </component>
</template>

<script lang="ts">
export const customProps = {
  origin: {
    default: 'top left',
    type: String
  },
  styles: {
    default: () => {
      return {
        animationFillMode: 'both',
        animationTimingFunction: 'cubic-bezier(.25,.8,.50,1)'
      }
    },
    type: Object
  }
}

export default {
  inheritAttrs: false
}
</script>

<script setup lang="ts">
import {
  buildComponentType,
  buildEmits,
  buildHooks,
  buildProps,
  buildTag
} from '../../composable'

const props = defineProps(buildProps(customProps))
const emit = defineEmits(buildEmits())

const componentType = buildComponentType(props)
const tag = buildTag(props)
const hooks = buildHooks(props, emit)

</script>

<style>
  @keyframes scale-in {
    from {
      opacity: 0;
      transform: scale(0);
    }

    to { opacity: 1; }
  }

  @keyframes scale-out {
    from { opacity: 1; }

    to {
      opacity: 0;
      transform: scale(0);
    }
  }

  .scale-in { animation-name: scale-in; }
  .scale-out { animation-name: scale-out; }
  .scale-move { transition: transform .3s cubic-bezier(.25, .8, .5, 1); }
</style>
