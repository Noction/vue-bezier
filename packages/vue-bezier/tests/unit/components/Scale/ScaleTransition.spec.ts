import ScaleTransition from '../../../../src/components/Scale/ScaleTransition.vue'
import { mount } from '@vue/test-utils'

describe('components / Scale / ScaleTransition', () => {
  it('When mounting component with default props, it should work fine', () => {
    const wrapper = mount(ScaleTransition)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('When mounting component with group enabled prop, it should work fine', () => {
    const wrapper = mount(ScaleTransition, { props: { group: true } })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
