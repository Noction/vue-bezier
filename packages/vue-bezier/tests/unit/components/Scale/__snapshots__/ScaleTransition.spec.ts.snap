// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components / Scale / ScaleTransition > When mounting component with default props, it should work fine 1`] = `"<transition-stub name="noc-scale" appear="false" persisted="false" css="true" tag="span"></transition-stub>"`;

exports[`components / Scale / ScaleTransition > When mounting component with group enabled prop, it should work fine 1`] = `"<transition-group-stub tag="span" name="noc-scale" appear="false" persisted="false" css="true"></transition-group-stub>"`;
